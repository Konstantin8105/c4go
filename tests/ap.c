#include "tests.h"
#include <stdio.h>

// input argument - C-pointer
void a(int* v1) { printf("a: %d\n", *v1); }

// input argument - C-array
void b(int v1[], int size)
{
    for (size--; size >= 0; size--) {
        printf("b: %d %d\n", size, v1[size]);
    }
}

long get()
{
    return (long)(0);
}

double global;

double* get_value()
{
    return &global;
}

int main()
{
    plan(4);

    diag("value");
    int i1 = 42;
    a(&i1);
    b(&i1, 1);

    diag("C-array");
    int i2[] = { 11, 22 };
    a(i2);
    b(i2, 2);

    diag("C-pointer from value");
    int* i3 = &i1;
    a(i3);
    b(i3, 1);

    diag("C-pointer from array");
    int* i4 = i2;
    a(i4);
    b(i4, 2);

    diag("C-pointer from array");
    int* i5 = &i2[1];
    a(i5);
    b(i5, 1);

    diag("pointer arithmetic 1");
    i5 = &i2[0];
    int* i6 = i5 + 1;
    a(i6);
    b(i6, 1);

    diag("pointer arithmetic 2");
    int val = 2 - 2;
    int* i7 = 1 + (1 - 1) + val + 0 * (100 - 2) + i5 + 0 - 0 * 0;
    a(i7);
    b(i7, 1);

    diag("pointer arithmetic 3");
    int* i8 = i5 + 1 + 0;
    a(i8);
    b(i8, 1);

    diag("pointer arithmetic 4");
    int i9[] = { *i3, *(i3 + 1) };
    a(i9);
    b(i9, 1);

    diag("pointer arithmetic 5");
    int* i10 = 1 + 0 + i5 + 5 * get() + get() + (12 + 3) * get();
    a(i10);
    b(i10, 1);

    diag("pointer arithmetic 6");
    int* i11 = 1 + 0 + i5 + 5 * get() + get() - (12 + 3) * get();
    a(i11);
    b(i11, 1);

    diag("pointer from function");
    global = 42;
    double* i12 = get_value();
    is_eq(*i12, global);

    diag("pointer for some type");
    typedef int NUMBER;
    typedef int* POINTER;
    NUMBER num;
    POINTER pnt;
    int num_value = 56;
    num = num_value;
    pnt = &num;
    is_eq(*pnt, num_value);
    *pnt = 123;
    is_eq(*pnt, 123);
    is_eq(num, 123);

    done_testing();
}
