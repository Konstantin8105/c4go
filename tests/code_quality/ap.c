#include <stdio.h>

// input argument - C-pointer
void a(int *v1) { printf("a: %d\n",*v1); }

// input argument - C-array
void b(int v1[], int size) {
	for (size-- ; size >= 0 ; size-- ) { printf("b: %d %d\n", size,  v1[size]); }
}

long get() {
	return (long)(0);
}

int main() {
	// value
	int i1 = 42; a(&i1); b(&i1, 1);

	// C-array
	int i2[] = {11,22} ; a(i2); b(i2,2);

	// C-pointer from value
	int *i3 = &i1      ; a(i3); b(i3,1);

	// C-pointer from array
	int *i4 = i2       ; a(i4); b(i4,2);

	// C-pointer from array
	int *i5 = i2[1]    ; a(i5); b(i5,1);

	// pointer arithmetic
	int *i6 = i5 + 1   ; a(i6); b(i6,1);

	// pointer arithmetic
	int val = 2-2;
	int *i7 = 1 + (1 - 1) + val + 0*(100-2) + i5 + 0 - 0*0; a(i7); b(i7,1);
	
	// pointer arithmetic
	int *i8 = i5 + 1 + 0 ; a(i8); b(i8,1);
	
	// pointer arithmetic
	int i9[] = {*i3, *(i3+1)} ; a(i9); b(i9,1);

	// pointer arithmetic
	int *i10 = 1 + 0 + i5 + 5*get() + get() + (12 + 3)*get(); a(i10); b(i10,1);

	// pointer arithmetic
	int *i11 = 1 + 0 + i5 + 5*get() + get() - (12 + 3)*get(); a(i11); b(i11,1);

	return 0;
}
