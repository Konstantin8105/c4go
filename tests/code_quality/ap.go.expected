//
//	Package - transpiled by c4go
//
//	If you have found any issues, please raise an issue at:
//	https://github.com/Konstantin8105/c4go/
//

package code_quality

import "reflect"
import "runtime"
import "unsafe"
import "github.com/Konstantin8105/c4go/noarch"

// a - transpiled function from  C4GO/tests/code_quality/ap.c:4
func a(v1 []int32) {
	// input argument - C-pointer
	noarch.Printf([]byte("a: %d\n\x00"), v1[0])
}

// b - transpiled function from  C4GO/tests/code_quality/ap.c:7
func b(v1 []int32, size int32) {
	// input argument - C-array
	for size -= 1; size >= 0; size-- {
		noarch.Printf([]byte("b: %d %d\n\x00"), size, v1[size])
	}
}

// get - transpiled function from  C4GO/tests/code_quality/ap.c:14
func get() int32 {
	return 0
}

// main - transpiled function from  C4GO/tests/code_quality/ap.c:19
func main() {
	// value
	var i1 int32 = 42
	a(c4goUnsafeConvert_int32(&i1))
	b(c4goUnsafeConvert_int32(&i1), 1)
	// C-array
	var i2 []int32 = []int32{11, 22}
	a(i2)
	b(i2, 2)
	// C-pointer from value
	var i3 []int32 = c4goUnsafeConvert_int32(&i1)
	a(i3)
	b(i3, 1)
	// C-pointer from array
	var i4 []int32 = i2
	a(i4)
	b(i4, 2)
	// C-pointer from array
	var i5 []int32 = i2[1:]
	a(i5)
	b(i5, 1)
	// pointer arithmetic
	var i6 []int32 = i5[0+1:]
	a(i6)
	b(i6, 1)
	// pointer arithmetic
	var val int32 = 2 - 2
	var i7 []int32 = c4goPointerArithInt32Slice(c4goPointerArithInt32Slice(c4goPointerArithInt32Slice(i5, int(1+(1-1)+val+0*(100-2))), int(0)), int(-(0 * 0)))
	a(i7)
	b(i7, 1)
	// pointer arithmetic
	var i8 []int32 = i5[0+1+0:]
	a(i8)
	b(i8, 1)
	// pointer arithmetic
	var i9 []int32 = []int32{i3[0], i3[0+1]}
	a(i9)
	b(i9, 1)
	// pointer arithmetic
	var i10 []int32 = i5[1+0+0+5*get()+get()+(12+3)*get():]
	a(i10)
	b(i10, 1)
	// pointer arithmetic
	var i11 []int32 = c4goPointerArithInt32Slice(i5[1+0+0+5*get()+get():], int(-((12 + 3) * get())))
	a(i11)
	b(i11, 1)
	return
}

// c4goUnsafeConvert_int32 : created by c4go
func c4goUnsafeConvert_int32(c4go_name *int32) []int32 {
	return (*[1000000]int32)(unsafe.Pointer(c4go_name))[:]
}

// c4goPointerArithInt32Slice - function of pointer arithmetic. generated by c4go
func c4goPointerArithInt32Slice(slice []int32, position int) []int32 {
	if position < 0 {
		// invert sign
		position = -position

		// Example from: go101.org/article/unsafe.html
		// repair size of slice
		var hdr reflect.SliceHeader
		sliceLen := len(slice)
		hdr.Data = uintptr(unsafe.Pointer(&slice[0])) - (uintptr(position))*unsafe.Sizeof(slice[0])
		runtime.KeepAlive(&slice[0]) // needed!
		hdr.Len = sliceLen + int(position)
		hdr.Cap = hdr.Len
		slice = *((*[]int32)(unsafe.Pointer(&hdr)))
		return slice
	}
	// position >= 0:
	return slice[position:]
}
